File: pure_c/bufdither_c/color_reducer.c
#include "color_reducer.h"

static int downgrade4lookup[256];

static int downgrade(int a, int targetBitCount) {
    int maxv = ((1 << targetBitCount) - 1);
    // ((a / 255.f) * maxv) / maxv * 255.f
    return a * maxv / 255 * 255 / maxv;
}

static void init_lookups() {
    for(int i = 0; i < 256; ++i) {
        downgrade4lookup[i] = downgrade(i, 4);
    }
}

static int downgrade4444(int a, int c) {
    //return downgrade(a, 4);
    return downgrade4lookup[a];
}

void color_reducer_init(color_reducer * r) {
    init_lookups();
    r->downgr = &downgrade4444;
}


void color_reducer_to_closest(const color_reducer * r, const trgba * src, trgba * dest) {
    for (int i = 0; i < 4; ++i) {
        dest->rgba[i] = r->downgr(src->rgba[i], i);
    }
}

File: luajit/color_reducer.lua
ffi = require 'ffi'
bit = require 'bit'

-----

local M = {}

local ColorReducer = {}
M.ColorReducer = ColorReducer

local downgrade4lookup = ffi.new("int32_t[?]", 256)

local function downgrade4444(a, cNum)
    return downgrade4lookup[a]
end

local function downgrade(a, targetBitCount)
    local maxv = bit.lshift(1, targetBitCount) - 1;
    return a * maxv / 255 * 255 / maxv
end

local function initLookups()
    for i = 0, 255 do
        downgrade4lookup[i] = downgrade(i, 4)
    end
end



function ColorReducer.create()
    initLookups()
    local inst = {}
    setmetatable(inst, {__index = ColorReducer})
    inst.downgr = downgrade4444
    return inst    
end


function ColorReducer:reduceToClosest(rgba, destRGBA)
    for i = 0,3 do
        destRGBA[i] = self.downgr(rgba[i], i)
    end
end


------
return M
